src/core/processing/queue.ts

```diff
@@ -0,0 +1,265 @@
+      else this.startTimer(node.next.date + this.handlerTimeout - Date.now())
+    }
+    this.remove(node)
+  }
+
+  /**
+   * Removes an element from the queue. Calls subscribers if there is capacity
+   * after performing this operation.
+   *
+   * @param node drift to remove
+   */
+  private remove(node: Drift<E>): void {
+    // Connecting the links of `prev` and `next` removes `node`
+    if (this.head === node) this.head = node.next
+    else node.prev!.next = node.next
+    if (this.tail === node) this.tail = node.prev
+    else node.next!.prev = node.prev
+
+    // Mark this drift as no longer contained
+    node.date = -1
+
+    // Notify subscribers if there is capacity by now
+    const capacity = this.concurrency - --this.length
+    if (capacity > 0) {
+      this.subscribers.forEach((resolve) => resolve(capacity))
+      this.subscribers = []
+    }
+  }
+
+  /**
+   * Takes a source element and starts the task for it by calling the worker
+   * function. Then wraps this task into a drift. Also makes sure that the drift
+   * removes itself from the queue once it completes, and that the error handler
+   * is invoked if it fails (rejects).
+   *
+   * @param elem source element
+   * @param date date when this drift is created
+   */
+  private toDrift(elem: E, date: number): Drift<E> {
+    const node: Drift<E> = {
+      prev: null,
+      task: this.worker(elem)
+        .catch(async (err) => {
+          // Rethrow iff the drift was removed by a timeout before
+          if (node.date > 0) await this.catchError(err, elem)
+          else throw err
+        })
+        .finally(() => {
+          // Decay the node once the task completes (unless the drift was
+          // removed due to a timeout before)
+          if (node.date > 0) this.decay(node)
+        }),
+      next: null,
+      date,
+      elem,
+    }
+    return node
+  }
+
+  /**
+   * Starts a timer that fires off a timeout after the given period of time.
+   *
+   * @param ms number of milliseconds to wait before the timeout kicks in
+   */
+  private startTimer(ms = this.handlerTimeout): void {
+    if (ms < 1) setImmediate(() => this.timeout())
+    else this.timer = setTimeout(() => this.timeout(), ms).unref()
+  }
+
+  /**
+   * Performs a timeout event. This removes the head element as well as all
+   * subsequent drifts with the same date (added in the same millisecond).
+   *
+   * The timeout handler is called in sequence for every removed drift.
+   */
+  private timeout(): void {
+    // Rare cases of the event ordering might fire a timeout even though the
+    // head element has just decayed.
+    if (this.head === null) return
+    while (this.head.date === this.head.next?.date) {
+      this.catchTimeout(this.head.elem, this.head.task)
+      // No need to restart timer here, we'll modify head again anyway
+      this.remove(this.head)
+    }
+    this.catchTimeout(this.head.elem, this.head.task)
```

Comment:
If this throws, process crashes synchronously, no cleanup, no syncing update offset.

Reply:
True, there should've been another `try`-`catch` around

Reply:
What would you do in the `catch` clause?

Reply:
Logging the error, or maybe something else if you have a better idea

Reply:
The best I can come up with is stopping polling and crashing. Inferior to #1262.

Reply:
The case that a timeout handler throws is completely independent of the implementation we choose. #1262 always calls the regular bot error handler instead, which does not provide any safety net at all. We can of course resemble that behaviour here if we wanted to, but #1262 is not even capable of distinguishing these very different errors from others. However, given that none of the two PRs have a chance of being merged in the current state, we may as well skip comparing them further. (This PR has a large diff but solves the problems, and the other one doesn't have a large diff but ignores some problems.)